File: cinema.py
#!/usr/bin/env python3
"""
Celebration Cinema showtime scraper.

Mode 1 — movies by day (default):
    python cinema.py [--days fri sat sun ...] [--weeks N]

Mode 2 — showtimes for a specific movie:
    python cinema.py --movie "mortal kombat" [--days fri sat sun ...] [--weeks N]

Supported day names: today, tomorrow, mon, tue, wed, thu, fri, sat, sun
Omitting --days shows all days within the --weeks window (default: 1 week).
"""

import argparse
import json
import re
import sys
from collections import defaultdict
from datetime import datetime

import requests
from bs4 import BeautifulSoup

URL = "https://www.celebrationcinema.com/cinemas/celebration-cinema-crossroads"
HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/124.0.0.0 Safari/537.36"
    )
}

# Maps 3-letter day names (and aliases) to Python weekday integers (Mon=0)
DAY_NAME_TO_WEEKDAY = {
    "mon": 0, "tue": 1, "wed": 2, "thu": 3,
    "fri": 4, "sat": 5, "sun": 6,
}


# ---------------------------------------------------------------------------
# Fetching & parsing
# ---------------------------------------------------------------------------

def fetch_html(url: str) -> str:
    resp = requests.get(url, headers=HEADERS, timeout=30)
    resp.raise_for_status()
    return resp.text


def extract_ng_init(page_html: str) -> str:
    soup = BeautifulSoup(page_html, "html.parser")
    tag = soup.find(attrs={"ng-init": re.compile(r"init\s*\(")})
    if not tag:
        sys.exit("Could not find ng-init tag with init() call.")
    return tag["ng-init"]


def parse_init_object(ng_init: str) -> dict:
    """Extract and fully parse the first object argument to init(...)."""
    start = ng_init.index("{")

    depth = 0
    in_string = False
    escape = False
    end = start
    for i, ch in enumerate(ng_init[start:], start):
        if escape:
            escape = False
            continue
        if ch == "\\" and in_string:
            escape = True
            continue
        if ch == '"':
            in_string = not in_string
            continue
        if in_string:
            continue
        if ch == "{":
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0:
                end = i
                break

    obj_str = ng_init[start:end + 1]

    try:
        raw_obj = json.loads(obj_str)
    except json.JSONDecodeError as exc:
        sys.exit(f"Failed to JSON-parse the init object: {exc}")

    parsed = {}
    for key, value in raw_obj.items():
        if isinstance(value, str):
            try:
                parsed[key] = json.loads(value)
            except json.JSONDecodeError:
                parsed[key] = value
        else:
            parsed[key] = value

    return parsed


def fetch_data() -> dict:
    print(f"Fetching {URL} ...", file=sys.stderr)
    page_html = fetch_html(URL)
    ng_init = extract_ng_init(page_html)
    return parse_init_object(ng_init)


# ---------------------------------------------------------------------------
# Day resolution
# ---------------------------------------------------------------------------

def resolve_days(requested: list[str], calendar_dates: list[dict], weeks: int = 1) -> list[tuple[str, str]]:
    """
    Given a list of day name tokens and the calendarDates list from the API,
    return an ordered list of (iso_date, label) pairs for the matched days,
    in calendar order, within the given number of weeks from today.

    calendar_dates entries: {"Text": "Today", "Moment": "2026-05-14T00:00:00", "ID": 0}
    """
    from datetime import date, timedelta
    cutoff = (date.today() + timedelta(weeks=weeks)).isoformat()

    # Build lookup structures from calendarDates
    # iso_date -> label
    cal_by_date: dict[str, str] = {}
    for entry in calendar_dates:
        iso = entry["Moment"][:10]  # "2026-05-14"
        cal_by_date[iso] = entry["Text"]

    # Ordered list of iso dates available within the window
    available_dates = [d for d in sorted(cal_by_date.keys()) if d <= cutoff]
    # today is first available date
    today_iso = sorted(cal_by_date.keys())[0] if cal_by_date else None

    # Resolve requested tokens -> set of iso dates
    matched: dict[str, str] = {}  # iso -> label, preserving calendar order later

    for token in requested:
        token = token.lower()
        if token == "today":
            if today_iso:
                matched[today_iso] = cal_by_date[today_iso]
        elif token == "tomorrow":
            tomorrow_candidates = available_dates[1:2]
            for d in tomorrow_candidates:
                matched[d] = cal_by_date[d]
        elif token in DAY_NAME_TO_WEEKDAY:
            target_wd = DAY_NAME_TO_WEEKDAY[token]
            for iso in available_dates:
                dt = datetime.fromisoformat(iso)
                if dt.weekday() == target_wd:
                    matched[iso] = cal_by_date[iso]
                    break
        else:
            print(f"Warning: unknown day '{token}' (use today/tomorrow/mon/tue/wed/thu/fri/sat/sun)", file=sys.stderr)

    # Return in calendar order
    return [(iso, matched[iso]) for iso in available_dates if iso in matched]


def all_days(calendar_dates: list[dict], weeks: int = 1) -> list[tuple[str, str]]:
    """Return available (iso_date, label) pairs within the given number of weeks from today."""
    from datetime import date, timedelta
    cutoff = (date.today() + timedelta(weeks=weeks)).isoformat()
    pairs = []
    for entry in sorted(calendar_dates, key=lambda e: e["Moment"]):
        iso = entry["Moment"][:10]
        if iso <= cutoff:
            pairs.append((iso, entry["Text"]))
    return pairs


# ---------------------------------------------------------------------------
# Formatting helpers
# ---------------------------------------------------------------------------

def fmt_time(iso_datetime: str) -> str:
    """Convert ISO datetime string to '5:35 PM'."""
    dt = datetime.fromisoformat(iso_datetime)
    return dt.strftime("%-I:%M %p")


def fmt_day_label(iso_date: str, cal_label: str) -> str:
    """Return a friendly day heading like 'Friday, May 16  (Today)'."""
    dt = datetime.fromisoformat(iso_date)
    weekday = dt.strftime("%A")
    month_day = dt.strftime("%B %-d")
    # Append calendar label if it's not just a date string
    if cal_label in ("Today", "Tomorrow"):
        return f"{weekday}, {month_day}  ({cal_label})"
    return f"{weekday}, {month_day}"


# ---------------------------------------------------------------------------
# Mode 1: movies by day
# ---------------------------------------------------------------------------

def mode_days(data: list[dict], day_pairs: list[tuple[str, str]], hide_times: bool = False) -> None:
    # Build: iso_date -> list of (title, runtime, sorted_showtimes)
    for iso_date, cal_label in day_pairs:
        # Collect movies that have showtimes on this date
        movies_on_day = []
        for movie in data:
            title = movie.get("Title", "Unknown")
            showtimes_on_day = [
                s for s in movie.get("Showtime", [])
                if s.get("Date", "")[:10] == iso_date
            ]
            if not showtimes_on_day:
                continue
            runtime = showtimes_on_day[0].get("RunTime", "")
            showtimes_on_day.sort(key=lambda s: s["Showtime"])
            movies_on_day.append((title, runtime, showtimes_on_day))

        if not movies_on_day:
            continue

        movies_on_day.sort(key=lambda m: m[0])

        heading = fmt_day_label(iso_date, cal_label)
        print(heading)
        print("-" * len(heading))
        for title, runtime, showtimes in movies_on_day:
            title_line = f"  {title}"
            if runtime:
                title_line += f"  [{runtime}]"
            print(title_line)
            if not hide_times:
                times_str = "  ".join(
                    f"{fmt_time(s['Showtime'])} ({s.get('FormatCode', '?')})"
                    for s in showtimes
                )
                print(f"    {times_str}")
        print()


# ---------------------------------------------------------------------------
# Mode 2: showtimes by movie
# ---------------------------------------------------------------------------

def mode_movie(data: list[dict], query: str, day_pairs: list[tuple[str, str]], hide_times: bool = False) -> None:
    query_lower = query.lower()
    matches = [m for m in data if query_lower in m.get("Title", "").lower()]

    if not matches:
        print(f"No movies found matching '{query}'.")
        return

    day_set = {iso for iso, _ in day_pairs}
    day_label_map = {iso: label for iso, label in day_pairs}

    for movie in matches:
        title = movie.get("Title", "Unknown")
        all_showtimes = movie.get("Showtime", [])

        # Group showtimes by date, filtered to requested days
        by_date: dict[str, list] = defaultdict(list)
        for s in all_showtimes:
            d = s.get("Date", "")[:10]
            if d in day_set:
                by_date[d].append(s)

        if not by_date:
            continue

        # Get runtime from first showtime
        first = next(iter(s for s in all_showtimes if s.get("RunTime")), None)
        runtime = first.get("RunTime", "") if first else ""

        title_line = title
        if runtime:
            title_line += f"  [{runtime}]"
        print(title_line)

        for iso_date in sorted(by_date.keys()):
            cal_label = day_label_map.get(iso_date, "")
            heading = fmt_day_label(iso_date, cal_label)
            print(f"  {heading}")
            if not hide_times:
                showtimes = sorted(by_date[iso_date], key=lambda s: s["Showtime"])
                times_str = "  ".join(
                    f"{fmt_time(s['Showtime'])} ({s.get('FormatCode', '?')})"
                    for s in showtimes
                )
                print(f"    {times_str}")
        print()


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(
        description="Celebration Cinema showtime scraper.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=(
            "Examples:\n"
            "  python cinema.py\n"
            "  python cinema.py --days fri sat sun\n"
            "  python cinema.py --weeks 2\n"
            "  python cinema.py --movie 'mortal kombat'\n"
            "  python cinema.py --movie 'mortal kombat' --days fri sat sun\n"
            "  python cinema.py --movie 'mortal kombat' --weeks 2\n"
        ),
    )
    parser.add_argument(
        "--days",
        nargs="+",
        metavar="DAY",
        help="Days to show: today tomorrow mon tue wed thu fri sat sun",
    )
    parser.add_argument(
        "--no-times",
        action="store_true",
        help="Hide showtimes, listing only movie titles per day",
    )
    parser.add_argument(
        "--weeks",
        type=int,
        default=1,
        metavar="N",
        help="How many weeks out to look (default: 1)",
    )
    parser.add_argument(
        "--movie",
        metavar="TITLE",
        help="Movie title search (case-insensitive substring match)",
    )
    args = parser.parse_args()

    result = fetch_data()
    calendar_dates = result.get("calendarDates", [])
    movie_data = result.get("data", [])

    if not calendar_dates:
        sys.exit("No calendarDates found in page data.")
    if not movie_data:
        sys.exit("No movie data found in page data.")

    if args.days:
        day_pairs = resolve_days(args.days, calendar_dates, weeks=args.weeks)
        if not day_pairs:
            sys.exit("None of the requested days are available in the schedule.")
    else:
        day_pairs = all_days(calendar_dates, weeks=args.weeks)

    print()

    if args.movie:
        mode_movie(movie_data, args.movie, day_pairs, hide_times=args.no_times)
    else:
        mode_days(movie_data, day_pairs, hide_times=args.no_times)


if __name__ == "__main__":
    main()

File: cinema.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
	"sort"
	"strings"
	"time"

	htmlparse "golang.org/x/net/html"
)

const siteURL = "https://www.celebrationcinema.com/cinemas/celebration-cinema-crossroads"

var userAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"

var dayNameToWeekday = map[string]time.Weekday{
	"sun": time.Sunday,
	"mon": time.Monday,
	"tue": time.Tuesday,
	"wed": time.Wednesday,
	"thu": time.Thursday,
	"fri": time.Friday,
	"sat": time.Saturday,
}

// ---------------------------------------------------------------------------
// Data types
// ---------------------------------------------------------------------------

type CalendarDate struct {
	Text   string `json:"Text"`
	Moment string `json:"Moment"`
	ID     int    `json:"ID"`
}

type Showtime struct {
	Date       string `json:"Date"`
	Showtime   string `json:"Showtime"`
	RunTime    string `json:"RunTime"`
	FormatCode string `json:"FormatCode"`
}

type Movie struct {
	Title    string     `json:"Title"`
	Showtime []Showtime `json:"Showtime"`
}

type InitData struct {
	CalendarDates []CalendarDate
	Movies        []Movie
}

// ---------------------------------------------------------------------------
// Fetching & parsing
// ---------------------------------------------------------------------------

func fetchHTML(url string) (string, error) {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("User-Agent", userAgent)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	return string(body), nil
}

var ngInitRe = regexp.MustCompile(`(?i)init\s*\(`)

func extractNgInit(pageHTML string) (string, error) {
	doc, err := htmlparse.Parse(strings.NewReader(pageHTML))
	if err != nil {
		return "", err
	}

	var found string
	var walk func(*htmlparse.Node)
	walk = func(n *htmlparse.Node) {
		if found != "" {
			return
		}
		if n.Type == htmlparse.ElementNode {
			for _, attr := range n.Attr {
				if attr.Key == "ng-init" && ngInitRe.MatchString(attr.Val) {
					found = attr.Val
					return
				}
			}
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			walk(c)
		}
	}
	walk(doc)

	if found == "" {
		return "", fmt.Errorf("could not find ng-init attribute with init() call")
	}
	return found, nil
}

// extractBracketObject extracts the first top-level {...} from s starting at
// the first '{', respecting nested braces and JSON strings.
func extractBracketObject(s string) (string, error) {
	start := strings.Index(s, "{")
	if start < 0 {
		return "", fmt.Errorf("no '{' found")
	}

	depth := 0
	inString := false
	escape := false

	for i := start; i < len(s); i++ {
		ch := s[i]
		if escape {
			escape = false
			continue
		}
		if ch == '\\' && inString {
			escape = true
			continue
		}
		if ch == '"' {
			inString = !inString
			continue
		}
		if inString {
			continue
		}
		if ch == '{' {
			depth++
		} else if ch == '}' {
			depth--
			if depth == 0 {
				return s[start : i+1], nil
			}
		}
	}
	return "", fmt.Errorf("unmatched '{' in string")
}

func parseInitData(ngInit string) (*InitData, error) {
	objStr, err := extractBracketObject(ngInit)
	if err != nil {
		return nil, fmt.Errorf("extracting init object: %w", err)
	}

	// The outer object has string values that are themselves JSON-encoded.
	var raw map[string]string
	if err := json.Unmarshal([]byte(objStr), &raw); err != nil {
		return nil, fmt.Errorf("unmarshaling init object: %w", err)
	}

	calJSON, ok := raw["calendarDates"]
	if !ok {
		return nil, fmt.Errorf("no calendarDates field")
	}
	dataJSON, ok := raw["data"]
	if !ok {
		return nil, fmt.Errorf("no data field")
	}

	var calDates []CalendarDate
	if err := json.Unmarshal([]byte(calJSON), &calDates); err != nil {
		return nil, fmt.Errorf("parsing calendarDates: %w", err)
	}

	var movies []Movie
	if err := json.Unmarshal([]byte(dataJSON), &movies); err != nil {
		return nil, fmt.Errorf("parsing data: %w", err)
	}

	return &InitData{CalendarDates: calDates, Movies: movies}, nil
}

func fetchData() (*InitData, error) {
	fmt.Fprintln(os.Stderr, "Fetching", siteURL, "...")
	pageHTML, err := fetchHTML(siteURL)
	if err != nil {
		return nil, err
	}
	ngInit, err := extractNgInit(pageHTML)
	if err != nil {
		return nil, err
	}
	return parseInitData(ngInit)
}

// ---------------------------------------------------------------------------
// Day resolution
// ---------------------------------------------------------------------------

type DayPair struct {
	ISO      string // "2026-05-16"
	CalLabel string // "Today", "Tomorrow", "Sat 5/16", etc.
}

func resolveDays(requested []string, calDates []CalendarDate, weeks int) ([]DayPair, error) {
	today := time.Now().Format("2006-01-02")
	cutoff := time.Now().AddDate(0, 0, weeks*7).Format("2006-01-02")

	calByDate := map[string]string{}
	var available []string
	for _, cd := range calDates {
		iso := cd.Moment[:10]
		calByDate[iso] = cd.Text
	}
	for iso := range calByDate {
		if iso >= today && iso <= cutoff {
			available = append(available, iso)
		}
	}
	sort.Strings(available)

	todayISO := ""
	if len(available) > 0 {
		todayISO = available[0]
	}

	matched := map[string]string{}
	for _, token := range requested {
		t := strings.ToLower(token)
		switch t {
		case "today":
			if todayISO != "" {
				matched[todayISO] = calByDate[todayISO]
			}
		case "tomorrow":
			if len(available) > 1 {
				matched[available[1]] = calByDate[available[1]]
			}
		default:
			wd, ok := dayNameToWeekday[t]
			if !ok {
				fmt.Fprintf(os.Stderr, "Warning: unknown day %q (use today/tomorrow/mon/tue/wed/thu/fri/sat/sun)\n", token)
				continue
			}
			for _, iso := range available {
				dt, _ := time.Parse("2006-01-02", iso)
				if dt.Weekday() == wd {
					matched[iso] = calByDate[iso]
					break
				}
			}
		}
	}

	var pairs []DayPair
	for _, iso := range available {
		if label, ok := matched[iso]; ok {
			pairs = append(pairs, DayPair{ISO: iso, CalLabel: label})
		}
	}
	return pairs, nil
}

func allDays(calDates []CalendarDate, weeks int) []DayPair {
	today := time.Now().Format("2006-01-02")
	cutoff := time.Now().AddDate(0, 0, weeks*7).Format("2006-01-02")

	var pairs []DayPair
	for _, cd := range calDates {
		iso := cd.Moment[:10]
		if iso >= today && iso <= cutoff {
			pairs = append(pairs, DayPair{ISO: iso, CalLabel: cd.Text})
		}
	}
	sort.Slice(pairs, func(i, j int) bool { return pairs[i].ISO < pairs[j].ISO })
	return pairs
}

// ---------------------------------------------------------------------------
// Formatting helpers
// ---------------------------------------------------------------------------

func fmtTime(iso string) string {
	// Showtimes are timezone-aware: "2026-05-16T17:35:00-04:00"
	t, err := time.Parse(time.RFC3339, iso)
	if err != nil {
		// fallback: strip offset and parse as local
		if len(iso) >= 19 {
			t, _ = time.ParseInLocation("2006-01-02T15:04:05", iso[:19], time.Local)
		}
	}
	return t.Format("3:04 PM")
}

func fmtDayLabel(iso, calLabel string) string {
	t, _ := time.Parse("2006-01-02", iso)
	base := t.Format("Monday, January 2")
	if calLabel == "Today" || calLabel == "Tomorrow" {
		return fmt.Sprintf("%s  (%s)", base, calLabel)
	}
	return base
}

func isFuture(iso string) bool {
	t, err := time.Parse(time.RFC3339, iso)
	if err != nil {
		return true // don't filter if we can't parse
	}
	return t.After(time.Now())
}

func shortLabel(iso, calLabel string) string {
	if calLabel == "Today" {
		return "today"
	}
	if calLabel == "Tomorrow" {
		return "tmrw"
	}
	t, _ := time.Parse("2006-01-02", iso)
	return strings.ToLower(t.Format("Mon"))
}

// ---------------------------------------------------------------------------
// Mode 1: movies by day
// ---------------------------------------------------------------------------

func modeDays(movies []Movie, pairs []DayPair, hideTimes bool) {
	for _, dp := range pairs {
		type entry struct {
			title     string
			runtime   string
			showtimes []Showtime
		}
		var dayMovies []entry

		for _, m := range movies {
			var sts []Showtime
			for _, s := range m.Showtime {
				if len(s.Date) >= 10 && s.Date[:10] == dp.ISO && isFuture(s.Showtime) {
					sts = append(sts, s)
				}
			}
			if len(sts) == 0 {
				continue
			}
			sort.Slice(sts, func(i, j int) bool { return sts[i].Showtime < sts[j].Showtime })
			runtime := ""
			if len(sts) > 0 {
				runtime = sts[0].RunTime
			}
			dayMovies = append(dayMovies, entry{m.Title, runtime, sts})
		}

		if len(dayMovies) == 0 {
			continue
		}
		sort.Slice(dayMovies, func(i, j int) bool { return dayMovies[i].title < dayMovies[j].title })

		heading := fmtDayLabel(dp.ISO, dp.CalLabel)
		fmt.Println(heading)
		fmt.Println(strings.Repeat("-", len(heading)))
		for _, e := range dayMovies {
			line := "  " + e.title
			if e.runtime != "" {
				line += "  [" + e.runtime + "]"
			}
			fmt.Println(line)
			if !hideTimes {
				var parts []string
				for _, s := range e.showtimes {
					parts = append(parts, fmt.Sprintf("%s (%s)", fmtTime(s.Showtime), s.FormatCode))
				}
				fmt.Println("    " + strings.Join(parts, "  "))
			}
		}
		fmt.Println()
	}
}

// ---------------------------------------------------------------------------
// Mode 2: showtimes by movie
// ---------------------------------------------------------------------------

func modeMovie(movies []Movie, query string, pairs []DayPair, hideTimes bool) {
	q := strings.ToLower(query)
	daySet := map[string]string{}
	for _, dp := range pairs {
		daySet[dp.ISO] = dp.CalLabel
	}

	var orderISO []string
	for _, dp := range pairs {
		orderISO = append(orderISO, dp.ISO)
	}

	found := false
	for _, m := range movies {
		if !strings.Contains(strings.ToLower(m.Title), q) {
			continue
		}

		byDate := map[string][]Showtime{}
		for _, s := range m.Showtime {
			d := ""
			if len(s.Date) >= 10 {
				d = s.Date[:10]
			}
			if _, ok := daySet[d]; ok && isFuture(s.Showtime) {
				byDate[d] = append(byDate[d], s)
			}
		}
		if len(byDate) == 0 {
			continue
		}
		found = true

		runtime := ""
		for _, s := range m.Showtime {
			if s.RunTime != "" {
				runtime = s.RunTime
				break
			}
		}

		title := m.Title
		if runtime != "" {
			title += "  [" + runtime + "]"
		}
		fmt.Println(title)

		for _, iso := range orderISO {
			sts, ok := byDate[iso]
			if !ok {
				continue
			}
			sort.Slice(sts, func(i, j int) bool { return sts[i].Showtime < sts[j].Showtime })
			fmt.Println("  " + fmtDayLabel(iso, daySet[iso]))
			if !hideTimes {
				var parts []string
				for _, s := range sts {
					parts = append(parts, fmt.Sprintf("%s (%s)", fmtTime(s.Showtime), s.FormatCode))
				}
				fmt.Println("    " + strings.Join(parts, "  "))
			}
		}
		fmt.Println()
	}

	if !found {
		fmt.Printf("No movies found matching %q.\n", query)
	}
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

var (
	daysFlag    string
	weeksFlag   int
	movieFlag   string
	noTimes     bool
	versionFlag bool
)

var Version = "dev"

func main() {
	flag.StringVar(&daysFlag, "days", "", "Comma separated days: today tomorrow mon tue wed thu fri sat sun")
	flag.IntVar(&weeksFlag, "weeks", 1, "How many weeks out to look (default: 1)")
	flag.StringVar(&movieFlag, "movie", "", "Movie title search (case-insensitive substring match)")
	flag.BoolVar(&noTimes, "no-times", false, "Hide showtimes, show titles only")
	flag.BoolVar(&versionFlag, "v", false, "Show version")
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, `Usage: cinema [options]

Examples:
  cinema --days "fri sat sun" --no-times
  cinema --days "fri sat sun"
  cinema --movie "mortal kombat" --days "fri sat sun"
  cinema --weeks 2

Options:
`)
		flag.PrintDefaults()
	}
	flag.Parse()

	if versionFlag {
		fmt.Println(Version)
		return
	}

	data, err := fetchData()
	if err != nil {
		fmt.Fprintln(os.Stderr, "Error:", err)
		os.Exit(1)
	}

	var pairs []DayPair
	if daysFlag != "" {
		// Accept both space and comma separated
		tokens := strings.FieldsFunc(daysFlag, func(r rune) bool { return r == ',' || r == ' ' })
		pairs, err = resolveDays(tokens, data.CalendarDates, weeksFlag)
		if err != nil || len(pairs) == 0 {
			fmt.Fprintln(os.Stderr, "None of the requested days are available in the schedule.")
			os.Exit(1)
		}
	} else {
		pairs = allDays(data.CalendarDates, weeksFlag)
	}

	fmt.Println()

	switch {
	case movieFlag != "":
		modeMovie(data.Movies, movieFlag, pairs, noTimes)
	default:
		modeDays(data.Movies, pairs, noTimes)
	}
}
